### PR TITLE
feat: add durable storage MVP

### PR DESCRIPTION
**Pull request by MasterPtato**
Resolves OGBE-97

### CODE REVIEW

**Comment by MasterPtato** [comment]
* **#267** <a href="https://app.graphite.dev/github/pr/rivet-gg/opengb/267?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @MasterPtato and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/runtime/durable.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+const ENCODER = new TextEncoder();
+const MEMORY_DURABLE_STORAGE: Map<string, {
+	durable: Durable;
+	storage: Map<string, any>;
+}> = new Map();
+
+export class Durable {
+	storage: DurableStorage;
+
+	constructor(public id: string) {
```
Comment (Blckbrry-Pi) [inline src/runtime/durable.ts:12]: If you don't want module authors to call the constructor by accident, make sure to make this a `private constructor`.

**Review by Blckbrry-Pi — CHANGES_REQUESTED** [review]
Looks great overall!

I don't know too much about durable storage and the idea behind it, but I think the Deno implementation is pretty severely lacking without persistence.

Additionally, most of the comments I left on the Deno impl also apply to the CF impl. 

Again, I apologize for advocating for more typescript complexity, but if instantiating a `Durable` it isn't disallowed by the type system, we NEED to explicitly document that. Module authors will definitely do that by accident.

Additionally, I'm not completely sure what the "point" of the durable storage is. Is that not what postgres and prisma are for?

File: tests/basic/modules/foo/scripts/call_self.ts
[the remark this answers is not in the document]
Comment (NathanFlurry) [reply tests/basic/modules/foo/scripts/call_self.ts:1]: +1 should be a separate module for testing core functionality e2e

**Comment by NathanFlurry** [comment]
@MasterPtato would you rather me wait until the next iteration is ready to go or review this now?

**Comment by MasterPtato** [comment]
@Blckbrry-Pi @NathanFlurry this was meant to be a draft sorry

File: src/runtime/context.ts [hunk 2]
```diff
@@ -145,9 +145,11 @@ export class Context<RegistryT, RegistryCamelT> {
...
 				throw cause;
 			} else {
-				const error = new RuntimeError("INTERNAL_ERROR", { cause });
-				error.enrich(this.runtime, this);
-				throw error;
+				throw cause;
+
+				// const error = new RuntimeError("INTERNAL_ERROR", { cause });
+				// error.enrich(this.runtime, this);
+				// throw error;
```
[the remark this answers is not in the document]
Comment (MasterPtato) [reply src/runtime/context.ts]: This didnt print the error message in the HTTP response so i returned the actual error instead. I just used it for debugging cf workers

File: src/runtime/durable.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		this.storage = new DurableStorage(id);
+	}
+
+	// Generics allow this function to match the type of whatever sub-class this function is called for
+	static async for<T extends typeof Durable>(this: T, name: string): Promise<InstanceType<T>> {
```
[the remark this answers is not in the document]
Comment (MasterPtato) [reply src/runtime/durable.ts]: This looks like too much black magic for a non-issue like this

File: src/runtime/durable.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+				durable,
+				storage: new Map(),
+			});
+
+			return durable as InstanceType<T>;
+		}
+	}
+}
+
+// In-memory storage implementation
```
[the remark this answers is not in the document]
Comment (MasterPtato) [reply src/runtime/durable.ts]: This is for the MVP and only used with `opengb dev`, the managed opengb version with cloudflare is persistent. We can add written file persistence later if needed

**Comment by NathanFlurry** [comment]
@MasterPtato merging this since it's blocking other things. let's make sure this gets cleaned up at some point soon.

**Comment by NathanFlurry** [comment]
### Merge activity

* **May 24, 11:41 PM EDT**: @NathanFlurry started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/rivet-gg/opengb/267).

**Pull request merged**